### DISCUSSION TITLE
Remote editing

### DISCUSSION THREAD

**Discussion by nvillar** [shorepine/tulipcc#129] [Ideas]
I love the concept of the Tulip as a standalone, distraction-free creative tool, where code and text editing can place on the device itself, with just a keyboard needed.

However when working on longer and more complex scripts it would be nice to have the option to use another device - like a desktop or laptop, together with the code editor of your choice - to write code, ideally with a mechanism to easily push and run on a Tulip CC hardware on the same network. 
 
I think, at the moment, doing this involves having a Desktop version of Tulip installed, opening and editing local files using an external editor, then using the Desktop Tulip to push to WORLD, then downloading and running on the Hardware Tulip. It's a fairly long cycle when you are trying things out and want fast, iterative changes. 

The Web REPL would kind of help, but as bwhitman points out it's buggy. It would also be great to be able to use something more fully featured like VS Code, to leverage all the code-completion and listing options, and things like Github Copilot.

Is there room for something like the ability to 'scp' files to/from a Tulip, plus the ability to remotely run specific scripts?

**Comment by bwhitman** [comment]
Absolutely room for this. For local dev like this i have a simple function in my `boot.py`:
```python
def grab(filename, run=True):
    tulip.url_save("http://localhost:8000/%s" %(filename), filename)
    if(run):
        execfile(filename)
```

And I run `python -m http.server` in my `~/Documents/tulipcc` folder on my computer. I edit with my normal computer text editor, do quick checks on a running Tulip Desktop instance, and when I want to test on hardware, I run `grab('file.py')` and it runs there. 

This doesn't let me go the other way -- from device to computer -- but that's more rare for me and I can use Tulip WORLD for that. 

I'm sure we can make this easier if you wanted it to be! Maybe look at `pyboard`? https://docs.micropython.org/en/latest/reference/pyboard.py.html Or `ampy`? https://github.com/scientifichackers/ampy

**Reply by andypiper** [reply]
ooh, this might be possible using `mpremote mount`... I'm literally barely hours in playing with this, but I might take a look at some stage. Would need tethering rather than over the network, though.

**Reply by andypiper** [reply]
Tried using `mpremote` from the Mac, but as soon as I run that, the Tulip device switches into raw REPL mode and becomes unresponsive. Need to think about how this could work.

**Comment by richmanaust** [comment]
I am on a mac and I use Thonny to upload sketches and as access to the micropython repl. I have used VScode to edit sketches quite successfully. I'm no expert, but have had some success with help from Claude Code.
